src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessStartInfo.cs

```diff
@@ -81,19 +81,50 @@ public string Arguments
                     _environmentVariables = new DictionaryWrapper(new Dictionary<string, string?>(
                         envVars.Count,
-                        CaseSensitiveEnvironmentVariables ? StringComparer.Ordinal : StringComparer.OrdinalIgnoreCase));
-#pragma warning restore 0429
+                        OperatingSystem.IsWindows() ? StringComparer.OrdinalIgnoreCase : StringComparer.Ordinal));
 
                     // Manual use of IDictionaryEnumerator instead of foreach to avoid DictionaryEntry box allocations.
                     IDictionaryEnumerator e = envVars.GetEnumerator();
                     Debug.Assert(!(e is IDisposable), "Environment.GetEnvironmentVariables should not be IDisposable.");
                     while (e.MoveNext())
                     {
                         DictionaryEntry entry = e.Entry;
-                        _environmentVariables.Add((string)entry.Key, (string?)entry.Value);
+
+                        string key = (string)entry.Key;
+                        string? value = (string?)entry.Value;
+
+                        if (OperatingSystem.IsWindows()) // it returns a const, the branch will be eliminated by JIT
+                        {
+                            if (_environmentVariables.TryAdd(key, value)) // fast path, 99.999% of the cases
+                            {
+                                continue;
+                            }
+
+                            // 1. Windows API used for setting env vars (SetEnvironmentVariableW) of the current process is case-insensitive:
+                            //      SetEnvironmentVariableW("A", "b");
+                            //      SetEnvironmentVariableW("a", "c"); // overwrites "b" with "c"
+                            //      Console.WriteLine(GetEnvironmentVariableW("A") + GetEnvironmentVariableW("a")); // Prints "cc"
+                            //    However, the CreateProcess API which accepts env vars for the new process, allows for case-sensitive duplicates.
+                            //    So it's possible that a Windows process can be started with DUPLICATE env vars.
+                            // 2. Environment.GetEnvironmentVariables uses GetEnvironmentStringsW which also accepts the duplicates.
+                            // 3. We don't want to change the existing API behaviour (this would be a breaking change).
+                            //    But we can allow for duplicates that have the same VALUE. Example from: https://github.com/dotnet/runtime/issues/42029
+                            //    "NPM_CONFIG_CACHE=^"
+                            //    "npm_config_cache=^"
+                            // 4. For duplicates that have different values, we don't know which one should be used and we just call
+                            //    Add() which is going to throw the same exception as it used to do (no breaking changes).
```

Comment:
If I understand this correctly, the motivation of the change is to avoid throwing in cases where `key` is already populated?

`_environmentVariables` is of type `DictionaryWrapper` which happens to implement `Add` [in terms of its indexer](https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Diagnostics.Process/src/System/Collections/Specialized/DictionaryWrapper.cs#L42), which has overwrite semantics and doesn't throw in that case.

So I think what this changes really is we're keeping the first duplicate instead of the last one (in the order of source dictionary enumeration).

The [stacktrace shared in the original issue](https://github.com/dotnet/msbuild/issues/5726#issuecomment-689770076) seems to indicate that the exception originates from a HashTable, making me suspect [this method is the culprit](https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs#L108).

Reply:
> If I understand this correctly, the motivation of the change is to avoid throwing in cases where key is already populated?

Not throwing, but only when it's populated with exactly the same value. Example:

```ini
(key, value)
"A", "b"
"a", "b"
// no throw, "b" == "b"

"A", "b"
"a", "c"
// throw (as we used to), because "b" != "c" and we don't know which one to use
```

Reply:
> which has overwrite semantics and doesn't throw in that case

I was thinking about it, but it could introduce silent errors for cases where the keys are the same, but values are different and so far it was throwing and giving clear error.

But I don't have a very strong opinion on this and perhaps I am just overthinking here.

Reply:
> The stacktrace shared in the original issue seems to indicate that the exception originates from a HashTable, making me suspect this method is the culprit.

This was my initial guess, but the `Hashtable` is created without specifying the comparer, so it's case sensitive by default.

https://github.com/dotnet/runtime/blob/31062e4ec6b248b709095e13947400b49d507e53/src/libraries/System.Private.CoreLib/src/System/Environment.Variables.Windows.cs#L82

A small repro to show that it works fine:

```cs
var set = new Hashtable();
set.Add("NPM_CONFIG_CACHE", "VALUE");
set.Add("npm_config_cache", "value");

Console.WriteLine(set.Count);
```

![obraz](https://user-images.githubusercontent.com/6011991/102216838-086cb000-3edc-11eb-9b1f-b1e18fdfe9ff.png)

I've one idea about it that I am going to share with you offline.

I've added tests that repro the failing case for all ways of getting env vars:

* `Environment.GetEnvironmentVariables()`
* `new ProcessStartInfo().Environment`
* `new ProcessStartInfo().EnvironmentVariables`